---
title: "Using PDclust"
author: "Tony Hui"
date: "`r Sys.Date()`"
output: 
  md_document:
    variant: markdown_github
---

```{r, echo = FALSE, include=FALSE}
library(dplyr)
library(pheatmap)
library(ggplot2)
library(PDclust, lib.loc = "~/dev")
knitr::opts_chunk$set(collapse = TRUE, comment = "#>", fig.width = 4, fig.height = 3, dpi = 150)
```

This package is based off the original PDclust publication \<insert citation here\>.

The workflow for doing PDclust is as follows:

## Step 0: Install

Easily install this package with `devtools`:

```
devtools::install_github("hui-tony-zk/PDclust")
```

Warning: this package requires the `tidyverse` suite of packages to work. I recommend you  `install.packages(tidyverse)` first. If you're unfamiliar with the `tidyverse` suite, I highly recommend you learn more about it.

## Step 1: Read in files

The input is a named list of data frames containing CpG calls for each single-cell. The four columns should be named as follows: `chr`, `start`, `end`, `meth` (for percent methylation).

There is a sample dataset included with this package that contains three single-cells and containing only chr22 to reduce file size.

```{r}
cpg_files
```

## Step 2: calculate pairwise dissimilarity with `create_pairwise_master()`

The second step is to do pairwise calculations. This results in a data.frame with the results

```{r}
cpg_files_pairwise <- create_pairwise_master(cpg_files)
cpg_files_pairwise
```

You can also parallelize this function over many cores, in cases with many samples. Keep in mind that there are ${n \choose 2} $ calculations, where `n` is the number of single-cells.

```{r}
create_pairwise_master(cpg_files, cores_to_use = 2)
```

You can include non-binary methylation calls (they're excluded by default as single-cells should be binary biologically)

```{r}
create_pairwise_master(cpg_files, digital = FALSE)
```

If you want to use your own metric, you can use this function to generate the pairwise joins

```{r}
create_pairwise_master(cpg_files, calcdiff = FALSE)
```

## Step 3: Convert to a symmetrical data frame using `convert_to_dissimilarity_matrix()`

This turns the data.frame into a clustering-ready matrix

```{r}
cpg_files_pairwise_matrix <- convert_to_dissimilarity_matrix(cpg_files_pairwise)
cpg_files_pairwise_matrix
```

## Step 4: Clustering

### A) Generating clusters with PDclust

This step clusters the single-cells together with `cluster_dissimilarity()`

```{r}
cluster_results <- cluster_dissimilarity(cpg_files_pairwise_matrix, num_clusters = 2)
```

You can plot the resulting cluster results:

```{r}
plot(cluster_results$hclust_obj)
```

... and view the clustering results

```{r}
cluster_results$cluster_assignments
```

... or use a fancy heatmap

```{r}
heatmap_pallete <- colorRampPalette(RColorBrewer::brewer.pal(8, name = "YlOrRd"))(21)

pheatmap(cpg_files_pairwise_matrix,
         cluster_rows = cluster_results$hclust_obj,
         cluster_cols = cluster_results$hclust_obj,
         treeheight_row = 0,
         border_color = NA,
         color = heatmap_pallete,
         show_colnames = F,
         annotation_col = cluster_results$cluster_assignments)
```


## B) Visualize the cluster separation:

This function projects the dissimilarities into 2D space (using `cmdscale`), and appends clustering results (if included) for easy coloring in `ggplot` later

```{r}
viz_df <- visualize_clusters(cpg_files_pairwise_matrix, cluster_results = cluster_results$cluster_assignments)
viz_df
```

```{r}
ggplot(viz_df, aes(V1, V2, color = cluster)) +
  geom_point()
```

